google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -0,0 +1,353 @@
+
+  /**
+   * The number of pending BulkWriter operations at which RecursiveDelete starts the next limit
+   * query to fetch descendants. By starting the query while there are pending operations, Firestore
+   * can improve BulkWriter throughput. This helps prevent BulkWriter from idling while Firestore
+   * fetches the next query.
+   */
+  public static final int MIN_PENDING_OPS = 1000;
+
+  private final FirestoreRpcContext<?> firestoreRpcContext;
+  private final BulkWriter writer;
+  @Nullable private final CollectionReference collectionReference;
+  @Nullable private final DocumentReference documentReference;
+
+  /** The number of deletes that failed with a permanent error. */
+  private int errorCount = 0;
+
+  /**
+   * The most recently thrown error. Used to populate the developer-facing error message when the
+   * recursive delete operation completes.
+   */
+  @Nullable private Throwable lastError;
+
+  /** Whether there are still documents to delete that still need to be fetched. */
+  private boolean documentsPending = true;
+
+  /** A deferred promise that resolves when the recursive delete operation is completed. */
+  private final SettableApiFuture<Void> completionFuture = SettableApiFuture.create();
+
+  /** Whether a query stream is currently in progress. Only one stream can be run at a time. */
+  private boolean streamInProgress = false;
+
+  /** Query limit to use when fetching all descendants. */
+  private int maxPendingOps = MAX_PENDING_OPS;
+
+  /**
+   * The number of pending BulkWriter operations at which RecursiveDelete starts the next limit
+   * query to fetch descendants.
+   */
+  private int minPendingOps = MIN_PENDING_OPS;
+
+  /**
+   * The last document snapshot returned by the stream. Used to set the startAfter() field in the
+   * subsequent stream.
+   */
+  @Nullable private DocumentSnapshot lastDocumentSnap;
+
+  /**
+   * The number of pending BulkWriter operations. Used to determine when the next query can be run.
+   */
+  private int pendingOperationsCount = 0;
+
+  RecursiveDelete(
+      FirestoreRpcContext<?> firestoreRpcContext,
+      BulkWriter writer,
+      CollectionReference reference) {
+    Preconditions.checkState(reference != null, "CollectionReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.collectionReference = reference;
+    this.documentReference = null;
+  }
+
+  RecursiveDelete(
+      FirestoreImpl firestoreRpcContext, BulkWriter writer, DocumentReference reference) {
+    Preconditions.checkState(reference != null, "DocumentReference cannot be null");
+    this.firestoreRpcContext = firestoreRpcContext;
+    this.writer = writer;
+    this.documentReference = reference;
+    this.collectionReference = null;
+  }
+
+  public ApiFuture<Void> run() {
+    Preconditions.checkState(
+        documentsPending, "The recursive delete operation has already been completed");
```

Comment:
This seems like it is not the correct check. We should only allow this call once. It should not be called again if it is already running, whether or not it is completed. You can either fix this check or remove it.